File: src/routing/orsProvider.js
/**
 * OpenRouteService routing provider.
 * https://openrouteservice.org
 *
 * Replaces OsrmProvider — the OSRM public demo server only serves a
 * "driving" profile, whereas ORS's free Standard API key covers driving,
 * cycling, and walking through the same endpoint (just a different profile
 * segment in the URL), which is what a "route on foot at Gatwick" test
 * actually needs.
 */
const OrsProvider = (() => {
  const BASE_URL   = "https://api.openrouteservice.org/v2/directions";
  const TIMEOUT_MS = 12000;

  const PROFILES = {
    driving: "driving-car",
    cycling: "cycling-regular",
    walking: "foot-walking",
  };

  /**
   * @param {{lat: number, lon: number}} start
   * @param {{lat: number, lon: number}} end
   * @param {"driving"|"cycling"|"walking"} [mode="driving"]
   * @returns {Promise<{geometry: object, distanceMeters: number, durationSeconds: number}|null>}
   */
  async function getRoute(start, end, mode) {
    const apiKey = CONFIG.ORS_API_KEY;
    if (!apiKey) {
      console.warn("OrsProvider: CONFIG.ORS_API_KEY not set — see config.js");
      return null;
    }

    const profile = PROFILES[mode] || PROFILES.driving;
    const url =
      `${BASE_URL}/${profile}?api_key=${encodeURIComponent(apiKey)}` +
      `&start=${start.lon},${start.lat}&end=${end.lon},${end.lat}`;

    const controller = new AbortController();
    const timer      = setTimeout(() => controller.abort(), TIMEOUT_MS);

    try {
      const res = await fetch(url, { signal: controller.signal });
      clearTimeout(timer);
      if (!res.ok) return null;

      const data    = await res.json();
      const feature = data.features && data.features[0];
      if (!feature) return null;

      const summary = feature.properties.summary;
      // Turn-by-turn steps, when ORS includes them — properties.segments is
      // one entry per requested leg (always one leg here, a single start/end
      // pair), each with its own .steps[] array. See maneuverTracker.js for
      // how these get turned into the guidance card's live instruction.
      // Defensive: an unexpected/older response shape just yields no steps,
      // not a crash — the guidance card falls back to its own geometry-only
      // detection when that happens.
      const segments = feature.properties.segments;
      const steps = (Array.isArray(segments) && segments[0] && Array.isArray(segments[0].steps))
        ? segments[0].steps
        : [];

      return {
        geometry:        feature.geometry, // GeoJSON LineString
        distanceMeters:  summary.distance,
        durationSeconds: summary.duration,
        steps,
      };
    } catch (err) {
      clearTimeout(timer);
      console.warn("OrsProvider: route request failed —", err.message);
      return null;
    }
  }

  return { getRoute };
})();

if (typeof module !== "undefined") module.exports = OrsProvider;

File: android/app/src/main/java/org/vectair/vcas/car/logic/OrsProvider.kt
package org.vectair.vcas.car.logic

import org.json.JSONObject

/**
 * OpenRouteService routing — a structural port of `src/routing/
 * orsProvider.js`'s response parsing, split into a pure, testable
 * `parseRouteResponse()` and the actual network call (`getRoute()`, not
 * unit-testable in this sandbox for the same reason `AdsbFiClient.kt`
 * isn't — no network access to a real ORS endpoint here). Same real ORS
 * Directions v2 GeoJSON response shape the JS original reads:
 * `features[0].properties.summary.{distance,duration}` and
 * `features[0].properties.segments[0].steps[]`, each step
 * `{distance, duration, type, instruction, name, way_points:[start,end]}`.
 *
 * `parseRouteResponse()` mirrors the JS original's defensive-by-field
 * reading exactly — a missing/malformed `segments`/`steps` degrades to an
 * empty steps list (not a crash), matching `ManeuverTracker`'s own doc
 * comment that this schema, while ORS's long-stable public one, was never
 * verified against a live response from a network-restricted sandbox.
 */
object OrsProvider {

    data class Step(
        val distanceMeters: Double,
        val durationSeconds: Double,
        val type: Int?,
        val instruction: String?,
        val name: String?,
        val wayPointStart: Int?,
        val wayPointEnd: Int?
    )

    /** `geometry` is [lon, lat] pairs — same convention Geo.kt/RouteGeometry.kt already use. */
    data class Route(
        val geometry: List<DoubleArray>,
        val distanceMeters: Double,
        val durationSeconds: Double,
        val steps: List<Step>
    )

    private val PROFILES = mapOf(
        "driving" to "driving-car",
        "cycling" to "cycling-regular",
        "walking" to "foot-walking"
    )

    fun profileFor(mode: String): String = PROFILES[mode] ?: PROFILES.getValue("driving")

    fun directionsUrl(baseUrl: String, apiKey: String, mode: String, startLat: Double, startLon: Double, endLat: Double, endLon: Double): String {
        val profile = profileFor(mode)
        return "$baseUrl/$profile?api_key=${java.net.URLEncoder.encode(apiKey, "UTF-8")}" +
            "&start=$startLon,$startLat&end=$endLon,$endLat"
    }

    fun parseRouteResponse(json: JSONObject): Route? {
        val features = json.optJSONArray("features") ?: return null
        if (features.length() == 0) return null
        val feature = features.optJSONObject(0) ?: return null

        val properties = feature.optJSONObject("properties") ?: return null
        val summary = properties.optJSONObject("summary") ?: return null
        val distanceMeters = summary.optDouble("distance", 0.0)
        val durationSeconds = summary.optDouble("duration", 0.0)

        val geometryObj = feature.optJSONObject("geometry")
        val coordsArr = geometryObj?.optJSONArray("coordinates")
        val geometry = ArrayList<DoubleArray>(coordsArr?.length() ?: 0)
        if (coordsArr != null) {
            for (i in 0 until coordsArr.length()) {
                val pair = coordsArr.optJSONArray(i) ?: continue
                if (pair.length() < 2) continue
                geometry.add(doubleArrayOf(pair.optDouble(0), pair.optDouble(1)))
            }
        }
        if (geometry.size < 2) return null

        val steps = ArrayList<Step>()
        val segments = properties.optJSONArray("segments")
        val firstSegment = segments?.optJSONObject(0)
        val stepsArr = firstSegment?.optJSONArray("steps")
        if (stepsArr != null) {
            for (i in 0 until stepsArr.length()) {
                val s = stepsArr.optJSONObject(i) ?: continue
                val wayPoints = s.optJSONArray("way_points")
                steps.add(
                    Step(
                        distanceMeters = s.optDouble("distance", 0.0),
                        durationSeconds = s.optDouble("duration", 0.0),
                        type = if (s.has("type") && !s.isNull("type")) s.optInt("type") else null,
                        instruction = if (s.has("instruction") && !s.isNull("instruction")) s.optString("instruction") else null,
                        name = if (s.has("name") && !s.isNull("name")) s.optString("name") else null,
                        wayPointStart = wayPoints?.optInt(0),
                        wayPointEnd = wayPoints?.optInt(1)
                    )
                )
            }
        }

        return Route(geometry, distanceMeters, durationSeconds, steps)
    }

    /**
     * Real network call — `HttpURLConnection`, same "plain framework API,
     * avoid a dependency this sandbox can't verify" discipline already
     * established for `AdsbFiClient.kt`. Must be called off the main
     * thread. Returns null on any failure (bad key, network error,
     * unparseable response) — the caller degrades gracefully, matching
     * the JS original's own `console.warn`-and-return-null behaviour.
     */
    fun getRoute(apiKey: String, mode: String, startLat: Double, startLon: Double, endLat: Double, endLon: Double): Route? {
        if (apiKey.isBlank()) return null
        val url = directionsUrl(BASE_URL, apiKey, mode, startLat, startLon, endLat, endLon)
        var connection: java.net.HttpURLConnection? = null
        return try {
            connection = (java.net.URL(url).openConnection() as java.net.HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                setRequestProperty("Accept", "application/json")
            }
            if (connection.responseCode != java.net.HttpURLConnection.HTTP_OK) return null
            val body = connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            parseRouteResponse(JSONObject(body))
        } catch (e: Exception) {
            null
        } finally {
            connection?.disconnect()
        }
    }

    private const val BASE_URL = "https://api.openrouteservice.org/v2/directions"
    private const val TIMEOUT_MS = 12000
}
